webauthn-security-key/libs/auth.js

```diff
@@ -0,0 +1,694 @@
+  return credentials.length > 0
+    ? authTypes.TWO_FACTOR
+    : authTypes.SINGLE_FACTOR;
+}
+
+function csrfCheck(req, res, next) {
+  if (req.header('X-Requested-With') != 'XMLHttpRequest') {
+    res.status(400).json({ error: 'Invalid access' });
+    return;
+  }
+  next();
+}
+
+function sessionCheck(req, res, next) {
+  if (req.session.name !== 'main') {
+    res.status(401).json({ error: 'Not authenticated' });
+    return;
+  }
+  next();
+}
+
+function getOrigin(userAgent) {
+  let origin = '';
+  if (userAgent.indexOf('okhttp') === 0) {
+    const octArray = process.env.ANDROID_SHA256HASH.split(':').map((h) =>
+      parseInt(h, 16)
+    );
+    const androidHash = base64url.encode(octArray);
+    origin = `android:apk-key-hash:${androidHash}`;
+  } else {
+    origin = process.env.ORIGIN;
+  }
+  return origin;
+}
+
+// ----------------------------------------------------------------------------
+// Database management and actions
+// ----------------------------------------------------------------------------
+
+function findUserByUsername(username) {
+  return db.get('users').find({ username }).value();
+}
+
+function createUser(user) {
+  db.get('users').push(user).write();
+}
+
+function updateUser(username, user) {
+  db.get('users').find({ username }).assign(user).write();
+}
+
+function updateCredentials(username, credentials) {
+  db.get('users').find({ username }).assign({ credentials }).write();
+}
+
+function createOrGetUser(username) {
+  let user = findUserByUsername(username);
+  // If the user doesn't exist, create it
+  if (!user) {
+    user = {
+      username,
+      id: base64url.encode(crypto.randomBytes(32)),
+      credentials: [],
+    };
+    createUser(user);
+  }
+  return user;
+}
+
+// ----------------------------------------------------------------------------
+// Session management
+// ----------------------------------------------------------------------------
+
+// Sign out
+router.get('/signout', (req, res) => {
+  // Remove the session
+  req.session.destroy();
+  // Redirect to `/`
+  res.redirect(307, '/');
+});
+
+/**
+ * Complete the authentication
+ *
+ * Input:
+ * req.session:
+ * {
+     username: String,
+     isPasswordCorrect: String
+ * }
+ *
+ * Response as JSON:
+ * {
+     msg: String, 
+     authStatus: String // One of authStatuses 
+   }
+ * or
+ * {
+     error: String, 
+   }
+ **/
+function completeAuthentication(req, res) {
+  // username and isPasswordCorrect come from the first session, 'auth', dedicated to authentication
+  const { username, isPasswordCorrect } = req.session;
+  if (!username || !isPasswordCorrect) {
+    res.status(401).json({ error: GENERIC_AUTH_ERROR_MESSAGE });
+    return;
+  }
+  // Terminate the 'auth' session and start the 'main' session
+  // Once the 'main' session is active, the user is considered fully authenticated
+  req.session.regenerate(function (err) {
+    req.session.name = 'main';
```

Comment:
WebAuthn session management question:
* 'auth' is the name of the session used for authentication
* 'main' is the name of the session once the user is authenticated
This naming may be confusing since the session name is 'auth' when the user is *not* authenticated - What's a better / commonly used naming?